### PR TITLE
Add media antivirus scanner

### PR DESCRIPTION
**Pull request by SBiOSoftWhare**

### CODE REVIEW

File: MatrixSDK/MXSession.m [hunk 1]
```diff
@@ -1576,7 +1580,9 @@ - (void)setAntivirusServerURL:(NSString *)antivirusServerURL
     // Update the media manager
     [mediaManager setAntivirusServerURL:antivirusServerURL];
     
-    // TODO: configure here a scan manager.
+    // Configure scan manager
```
Comment (giomfo) [inline MatrixSDK/MXSession.m]: You may mention here that the scan manager will be nil if the provided url is nil

File: MatrixSDK/JSONModels/MXEvent.h [hunk 2]
```diff
@@ -403,6 +403,19 @@ extern NSString *const kMXEventIdentifierKey;
...
 
+/**
+ Retrieve all the media URLs contained in the event.
+
+ @return All available media URLs.
+ */
+- (NSArray<NSString*>*)getMediaURLs;
+
+/**
+ Indicate if event content could be scan by `MXScanManager`.
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.h]: scanned

File: MatrixSDK/JSONModels/MXEvent.m [hunk 3]
```diff
@@ -521,6 +522,123 @@ - (NSComparisonResult)compareOriginServerTs:(MXEvent *)otherEvent
     return result;
 }
 
+- (NSArray<NSString*>*)getMediaURLs
+{
+    NSMutableArray<NSString*> *mediaURLs = [NSMutableArray new];
+    
+    if ([self.type isEqualToString:kMXEventTypeStringRoomMessage])
+    {
+        NSString *messageType = self.content[@"msgtype"];
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.m]: Can we use MXJSONSetString to avoid crash on bad formatted messages

File: MatrixSDK/JSONModels/MXEvent.m [hunk 4]
```diff
@@ -521,6 +522,123 @@ - (NSComparisonResult)compareOriginServerTs:(MXEvent *)otherEvent
...
+        
+        if ([messageType isEqualToString:kMXMessageTypeImage] || [messageType isEqualToString:kMXMessageTypeVideo])
+        {
+            NSString *mediaURL;
+            NSString *mediaThumbnailURL;
+            
+            if (self.isEncrypted)
+            {
+                mediaURL = self.content[@"file"][@"url"];
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.m]: MXJSONSetDictionary & MXJSONSetString

File: MatrixSDK/JSONModels/MXEvent.m [hunk 5]
```diff
@@ -521,6 +522,123 @@ - (NSComparisonResult)compareOriginServerTs:(MXEvent *)otherEvent
...
+                
+                NSDictionary *thubmnailFile = self.content[@"info"][@"thumbnail_file"];
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.m]: MXJSONSetDictionary

File: MatrixSDK/JSONModels/MXEvent.m [hunk 6]
```diff
@@ -521,6 +522,123 @@ - (NSComparisonResult)compareOriginServerTs:(MXEvent *)otherEvent
...
+                
+                if (thubmnailFile)
+                {
+                    mediaThumbnailURL = thubmnailFile[@"url"];
+                }
+            }
+            else
+            {
+                mediaURL = self.content[@"url"];
+                mediaThumbnailURL = self.content[@"info"][@"thumbnail_url"];
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.m]: etc...
Please do it everywhere it is required in this method.

File: MatrixSDK/JSONModels/MXEvent.m [hunk 7]
```diff
@@ -521,6 +522,123 @@ - (NSComparisonResult)compareOriginServerTs:(MXEvent *)otherEvent
...
+                NSDictionary *thubmnailFile = self.content[@"info"][@"thumbnail_file"];
+                
+                if (thubmnailFile)
+                {
+                    mediaThumbnailURL = thubmnailFile[@"url"];
+                }
+            }
+            else
+            {
+                mediaURL = self.content[@"url"];
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.m]: MXJSONSetString

File: MatrixSDK/JSONModels/MXEvent.m [hunk 8]
```diff
@@ -622,7 +740,29 @@ - (NSString *)claimedEd25519Key
     }
 }
 
+- (MXEncryptedContentFile*)getEncryptedThumbnaileFile
+{
+    MXEncryptedContentFile *encryptedContentFile;
+    
+    NSDictionary *contentInfo;
+    MXJSONModelSetDictionary(contentInfo, self.content[@"info"]);
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.m]: 👍

File: MatrixSDK/ContentScan/Data/Store/MXEventScanStore.h [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ The `MXEventScanStore` protocol defines an interface that must be implemented to manipulate event scans data.
+ */
+@protocol MXEventScanStore <NSObject>
+
+@property (nonatomic, weak, nullable) id<MXEventScanStoreDelegate> delegate;
+
+- (nullable MXEventScan*)findWithId:(nonnull NSString*)eventId;
+
+- (nonnull MXEventScan*)createOrUpdateWithId:(nonnull NSString*)eventId andMediaURLs:(nonnull NSArray<NSString*>*)mediaURLs;
```
Comment (manuroe) [inline MatrixSDK/ContentScan/Data/Store/MXEventScanStore.h]: We should document for example that `MXEventScan` must be created with the status `MXAntivirusScanStatusUnknown`.

File: MatrixSDK/JSONModels/MXEvent.m [hunk 10]
```diff
@@ -606,19 +608,20 @@ - (NSComparisonResult)compareOriginServerTs:(MXEvent *)otherEvent
         
         if (self.isEncrypted)
         {
-            mediaURL = self.content[@"file"][@"url"];
+            MXJSONModelSetString(mediaURL, self.content[@"file"][@"url"]);
             
-            NSDictionary *thubmnailFile = self.content[@"info"][@"thumbnail_file"];
+            NSDictionary *thubmnailFile;
+            MXJSONModelSetDictionary(thubmnailFile, self.content[@"info"][@"thumbnail_file"]);
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.m]: MXJSONModelSetDictionary

File: MatrixSDK/JSONModels/MXEvent.m [hunk 11]
```diff
@@ -606,19 +608,20 @@ - (NSComparisonResult)compareOriginServerTs:(MXEvent *)otherEvent
...
-                mediaThumbnailURL = thubmnailFile[@"url"];
+                MXJSONModelSetDictionary(mediaThumbnailURL, thubmnailFile[@"url"]);
             }
         }
         else
         {
-            mediaURL = self.content[@"url"];
-            mediaThumbnailURL = self.content[@"info"][@"thumbnail_url"];
+            MXJSONModelSetString(mediaURL, self.content[@"url"]);
+            MXJSONModelSetString(mediaThumbnailURL, self.content[@"info"][@"thumbnail_url"]);
```
Comment (manuroe) [inline MatrixSDK/JSONModels/MXEvent.m]: MXJSONModelSetDictionary

**Pull request merged**